In [1]:
import pandas as pd
import plotly.express as px
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn import linear_model



pd.set_option('display.max_columns', None) 
pd.set_option('display.max_rows', None)

In [2]:
df = pd.read_csv(r'E:\\ROLL\DataScience\\Projetos\\ds-salary-streamlit\\dataset\\updated_ds_jobs_2021.csv')
df.shape

(722, 42)

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 722 entries, 0 to 721
Data columns (total 42 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   job_title                  722 non-null    object 
 1   seniority                  722 non-null    object 
 2   job_title_sim              722 non-null    object 
 3   job_description            722 non-null    object 
 4   python                     722 non-null    int64  
 5   spark                      722 non-null    int64  
 6   aws                        722 non-null    int64  
 7   excel                      722 non-null    int64  
 8   sql                        722 non-null    int64  
 9   sas                        722 non-null    int64  
 10  keras                      722 non-null    int64  
 11  pytorch                    722 non-null    int64  
 12  scikit                     722 non-null    int64  
 13  tensor                     722 non-null    int64  

In [4]:
df.sample(5)

job_title seniority    job_title_sim  \
1             Healthcare Data Scientist        na   Data Scientist   
600         Medical Lab Scientist - MLT        na  Other Scientist   
442                      Data Scientist        na   Data Scientist   
522     Scientist - Analytical Services        na  Other Scientist   
619  Senior Scientist - Biostatistician        sr  Other Scientist   

                                       job_description  python  spark  aws  \
1    What You Will Do:  I. General Summary  The Hea...       1      0    0   
600  Position Type: Regular Scheduled Hours per 2 w...       0      0    0   
442  Description  Job Description:   The Defense gr...       1      0    1   
522  British American Tobacco Reynolds American Inc...       0      0    0   
619  British American Tobacco Reynolds American Inc...       0      0    0   

     excel  sql  sas  keras  pytorch  scikit  tensor  hadoop  tableau  bi  \
1        0    0    0      0        0       0       0       0        0   0   
600      0    0    0      0        0       0       0       0        0   0   
442      0    0    0      0        1       0       1       0        0   0   
522      1    0    0      0        0       0       0       0        0   0   
619      1    0    0      0        0       0       0       0        0   0   

     flink  mongo  google_an higher_degree  higher_degree_needed  \
1        0      0          0        Master                     1   
600      0      0          0           NaN                     0   
442      0      0          0        Master                     1   
522      0      0          0           NaN                     0   
619      0      0          0        Master                     1   

     avg_salary_hour  avg_salary_month  avg_salary_year  employer_provided  \
1              42.07            7292.0          87506.0                  0   
600            51.50            8927.0         107120.0                  0   
442            48.80            8459.0         101504.0                  0   
522            47.84            8292.0          99507.0                  0   
619            38.70            6708.0          80496.0                  0   

       job_location job_location_city                                company  \
1          Maryland         Linthicum  University of Maryland Medical System   
600        Nebraska             Omaha            Catholic Health Initiatives   
442        Virginia       Springfield                                 Leidos   
522  North Carolina     Winston-Salem                      Reynolds American   
619  North Carolina     Winston-Salem                      Reynolds American   

    company_headquarters company_headquarters_city  same_place  rating  \
1               Maryland                 Baltimore           0     3.4   
600             Colorado                 Englewood           0     3.2   
442             Virginia                    Reston           0     3.4   
522       North Carolina             Winston-Salem           1     3.1   
619       North Carolina             Winston-Salem           1     3.1   

    number_of_workers            revenue(USD) founded       type_of_ownership  \
1              10000+  $2 to $5 billion (USD)    1984      Other Organization   
600            10000+      $10+ billion (USD)    1996  Nonprofit Organization   
442            10000+      $10+ billion (USD)    1969        Company - Public   
522      5001 - 10000      $10+ billion (USD)    1875       Company - Private   
619      5001 - 10000      $10+ billion (USD)    1875       Company - Private   

                             industry               sector  age  \
1    Health Care Services & Hospitals          Health Care   37   
600  Health Care Services & Hospitals          Health Care   25   
442               Aerospace & Defense  Aerospace & Defense   52   
522   Consumer Products Manufacturing        Manufacturing  146   
619   Consumer Products Manufacturing        Manufa

In [5]:
df.describe()

python       spark         aws      excel         sql         sas  \
count  722.000000  722.000000  722.000000  722.00000  722.000000  722.000000   
mean     0.536011    0.224377    0.236842    0.51662    0.515235    0.091413   
std      0.499047    0.417460    0.425439    0.50007    0.500114    0.288395   
min      0.000000    0.000000    0.000000    0.00000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.00000    0.000000    0.000000   
50%      1.000000    0.000000    0.000000    1.00000    1.000000    0.000000   
75%      1.000000    0.000000    0.000000    1.00000    1.000000    0.000000   
max      1.000000    1.000000    1.000000    1.00000    1.000000    1.000000   

            keras     pytorch      scikit      tensor      hadoop     tableau  \
count  722.000000  722.000000  722.000000  722.000000  722.000000  722.000000   
mean     0.040166    0.054017    0.074792    0.098338    0.163435    0.198061   
std      0.196485    0.226207    0.263238    0.297977    0.370018    0.398815   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
50%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
75%      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

               bi       flink       mongo   google_an  higher_degree_needed  \
count  722.000000  722.000000  722.000000  722.000000            722.000000   
mean     0.074792    0.013850    0.048476    0.019391              0.480609   
std      0.263238    0.116951    0.214920    0.137989              0.499970   
min      0.000000    0.000000    0.000000    0.000000              0.000000   
25%      0.000000    0.000000    0.000000    0.000000              0.000000   
50%      0.000000    0.000000    0.000000    0.000000              0.000000   
75%      0.000000    0.000000    0.000000    0.000000              1.000000   
max      1.000000    1.000000    1.000000    1.000000              1.000000   

       avg_salary_hour  avg_salary_month  avg_salary_year  employer_provided  \
count       722.000000        722.000000       722.000000         722.000000   
mean         49.577119       8593.382271    103120.416898           0.009695   
std          17.674990       3063.662482     36763.988856           0.098054   
min           7.450000       1291.000000     15496.000000           0.000000   
25%          36.780000       6375.000000     76502.000000           0.000000   
50%          47.480000       8230.000000     98758.500000           0.000000   
75%          59.560000      10323.500000    123885.000000           0.000000   
max         122.120000      21168.000000    254010.000000           1.000000   

       same_place      rating         age  number_competitors  
count  722.000000  722.000000  722.000000          722.000000  
mean     0.566482    3.669806   48.202216            1.080332  
std      0.495904    0.598127   53.754082            1.392142  
min      0.000000    0.000000    0.000000            0.000000  
25%      0.000000    3.300000   13.000000            0.000000  
50%      1.000000    3.700000   26.000000            0.000000  
75%      1.000000    4.000000   60.000000            3.000000  
max      1.000000    5.000000  277.000000            4.000000

In [6]:
df.describe(exclude="number")

job_title seniority   job_title_sim  \
count              722       722             722   
unique             254         3               9   
top     Data Scientist        na  Data Scientist   
freq               128       504             310   

                                          job_description higher_degree  \
count                                                 722           347   
unique                                                451             2   
top     As we strive to make a better day for our gues...        Master   
freq                                                    4           247   

       job_location job_location_city                 company  \
count           722               722                     722   
unique           37               193                     335   
top      California          New York  Takeda Pharmaceuticals   
freq            150                54                      14   

       company_headquarters company_headquarters_city number_of_workers  \
count                   722                       722               722   
unique                   46                       188                 7   
top              California                  New York       1001 - 5000   
freq                    161                        52               148   

                    revenue(USD)  founded  type_of_ownership  \
count                        722      722                722   
unique                        13      102                  9   
top     Unknown / Non-Applicable  No Data  Company - Private   
freq                         191       39                393   

                         industry                  sector competitors  
count                         722                     722         722  
unique                         60                      25         128  
top     Biotech & Pharmaceuticals  Information Technology     No Data  
freq                          112                     173         441

In [7]:
df.groupby(['seniority','type_of_ownership'])['avg_salary_month'].mean()

seniority  type_of_ownership             
jr         Company - Private                  4708.000000
           Nonprofit Organization             6521.000000
na         College / University               8115.000000
           Company - Private                  7971.144828
           Company - Public                   8302.941176
           Government                         6944.200000
           Hospital                           8318.111111
           Nonprofit Organization             7182.761905
           Other Organization                 5292.000000
           School / School District           6479.000000
           Subsidiary or Business Segment     7415.136364
sr         College / University              10334.000000
           Company - Private                 10160.950980
           Company - Public                  10884.684932
           Government                         8291.400000
           Hospital                           7041.000000
           Nonprofit Organizat

In [8]:
df.isnull().sum()

job_title                      0
seniority                      0
job_title_sim                  0
job_description                0
python                         0
spark                          0
aws                            0
excel                          0
sql                            0
sas                            0
keras                          0
pytorch                        0
scikit                         0
tensor                         0
hadoop                         0
tableau                        0
bi                             0
flink                          0
mongo                          0
google_an                      0
higher_degree                375
higher_degree_needed           0
avg_salary_hour                0
avg_salary_month               0
avg_salary_year                0
employer_provided              0
job_location                   0
job_location_city              0
company                        0
company_headquarters           0
company_he

To make things simpler, we will only use the skills columns and use the avg_salary_month as our target variable

In [9]:
manter = (["python", "spark", "aws", "excel", "sql", "sas", "keras", "pytorch", "scikit", "tensor", "hadoop", "tableau", "bi", "flink", "mongo", "google_an", "avg_salary_month"])

df = df[manter]
df.sample(4)

python  spark  aws  excel  sql  sas  keras  pytorch  scikit  tensor  \
236       1      0    0      1    1    0      0        0       0       0   
452       1      0    0      0    1    0      0        0       0       0   
30        1      0    0      0    0    0      0        0       0       0   
71        0      0    0      1    0    0      0        0       0       0   

     hadoop  tableau  bi  flink  mongo  google_an  avg_salary_month  
236       0        0   0      0      0          0            6334.0  
452       0        1   1      0      0          0            4084.0  
30        0        0   0      0      0          0            6375.0  
71        0        0   0      0      0          0            6334.0

In [10]:

heatmap = px.imshow(df.corr(), color_continuous_scale="BuPu", aspect="auto", labels=dict(x="Grafico de Correlacao"))
heatmap.update_xaxes(side="top")
heatmap.show()

In [11]:
scatter = px.scatter(df["avg_salary_month"])
scatter.show()

Padronizando e Normalizando os dados

In [12]:
#Padronizacao dos dados -> Media = 0, Desvio = 1

df_padronizado = pd.DataFrame(StandardScaler().fit_transform(df), columns=df.columns, index=df.index)

df_padronizado.sample(3)

python     spark       aws     excel       sql      sas     keras  \
650  0.930394  1.859245 -0.557086 -1.033812 -1.030950 -0.31719 -0.204565   
534 -1.074813 -0.537853 -0.557086  0.967294  0.969979 -0.31719 -0.204565   
532 -1.074813 -0.537853 -0.557086 -1.033812 -1.030950 -0.31719 -0.204565   

      pytorch    scikit    tensor  hadoop   tableau        bi     flink  \
650 -0.238958 -0.284321 -0.330247  -0.442 -0.496968 -0.284321 -0.118511   
534 -0.238958 -0.284321 -0.330247  -0.442  2.012201 -0.284321 -0.118511   
532 -0.238958 -0.284321 -0.330247  -0.442 -0.496968 -0.284321 -0.118511   

        mongo  google_an  avg_salary_month  
650 -0.225713   -0.14062         -0.942461  
534 -0.225713   -0.14062         -1.132561  
532 -0.225713   -0.14062         -0.888240

In [13]:
#Normalizacao dos dados -> Valores entre 0 e 1

df_normalizado =  pd.DataFrame(MinMaxScaler().fit_transform(df), columns=df.columns, index=df.index)

df_normalizado.sample(3)

python  spark  aws  excel  sql  sas  keras  pytorch  scikit  tensor  \
436     1.0    1.0  0.0    1.0  1.0  0.0    0.0      1.0     1.0     1.0   
113     0.0    0.0  0.0    0.0  0.0  0.0    0.0      0.0     0.0     0.0   
85      0.0    0.0  0.0    0.0  1.0  0.0    0.0      0.0     0.0     0.0   

     hadoop  tableau   bi  flink  mongo  google_an  avg_salary_month  
436     1.0      0.0  0.0    0.0    0.0        0.0          0.383609  
113     0.0      0.0  0.0    0.0    0.0        0.0          0.272526  
85      1.0      0.0  0.0    0.0    0.0        0.0          0.517835

Separando os dados em um dataframe com apenas a variavel Target e outro com apenas as variaves preditoras, e separando os dados em treino e teste com o dataframe original


In [14]:
#gerando um dataframe com as variaves preditoras e sem a variavel target
var_preditoras = df.drop("avg_salary_month", axis = 1)
var_preditoras.shape

(722, 16)

In [15]:
#gerando um dataframe com a variavel target
var_target = df[["avg_salary_month"]]
var_target.shape

(722, 1)

In [16]:
#Separando os dados em 70% para treino e 30% para testes

X_train, X_test, y_train, y_test = train_test_split(var_preditoras, var_target, test_size=0.3, random_state=10)


Aplicando Linear Regression

In [17]:
linear_regression = linear_model.LinearRegression()
linear_regression.fit(X_train, y_train)

LinearRegression()

Observado os resultados

Mean Absolute Error (MAE): The average of absolute errors, quanto menor, melhor

Mean Squared Error (MSE): The average of squared errors, quanto menor, melhor


In [18]:
resultados = linear_regression.predict(X_test)

#testando o MAE
mae_score = mean_absolute_error(y_test, resultados)
print("Mean Absolute Error: " , mae_score)

#testando o MSE
mse_score = mean_squared_error(y_test, resultados)
print("Mean Squared Error: " , mse_score)

Mean Absolute Error:  2297.946475094325
Mean Squared Error:  8444570.685387319


Repetindo todo o processo com o dataframe padronizado

In [19]:
#gerando um dataframe com as variaves preditoras e sem a variavel target
var_preditoras = df_padronizado.drop("avg_salary_month", axis = 1)

#gerando um dataframe com a variavel target
var_target = df_padronizado[["avg_salary_month"]]

In [20]:
#Separando os dados em 70% para treino e 30% para testes

X_train, X_test, y_train, y_test = train_test_split(var_preditoras, var_target, test_size=0.3, random_state=10)

In [21]:
linear_regression = linear_model.LinearRegression()
linear_regression.fit(X_train, y_train)

LinearRegression()

In [22]:
resultados = linear_regression.predict(X_test)

#testando o MAE
mae_score = mean_absolute_error(y_test, resultados)
print("Mean Absolute Error: " , mae_score)

#testando o MSE
mse_score = mean_squared_error(y_test, resultados)
print("Mean Squared Error: " , mse_score)

Mean Absolute Error:  0.7505851312298671
Mean Squared Error:  0.9009437397819883


Repetindo todo o processo com o dataframe normalizado

In [23]:
#gerando um dataframe com as variaves preditoras e sem a variavel target
var_preditoras = df_normalizado.drop("avg_salary_month", axis = 1)

#gerando um dataframe com a variavel target
var_target = df_normalizado[["avg_salary_month"]]


In [24]:
#Separando os dados em 70% para treino e 30% para testes

X_train, X_test, y_train, y_test = train_test_split(var_preditoras, var_target, test_size=0.3, random_state=10)

In [25]:
linear_regression = linear_model.LinearRegression()
linear_regression.fit(X_train, y_train)

LinearRegression()

In [26]:
resultados = linear_regression.predict(X_test)

#testando o MAE
mae_score = mean_absolute_error(y_test, resultados)
print("Mean Absolute Error: " , mae_score)

#testando o MSE
mse_score = mean_squared_error(y_test, resultados)
print("Mean Squared Error: " , mse_score)

Mean Absolute Error:  0.11560831489129772
Mean Squared Error:  0.021373512517759537


Podemos perceber os seguintes resultados:

DF Original = MAE: 2297.946475094325
&nbsp;
 , MSE: 8444570.685387319 

DF Padronizado = MAE: 0.7505851312298671
&nbsp;
 , MSE: 0.9009437397819883

DF Normalizado = MAE: 0.11560831489129772
&nbsp;
 , MSE: 0.021373512517759537


Quanto menor o MAE e o MSE, melhor o resultado, portanto, neste caso o Dataframe normalizado obteve o melhor resultado para o LinearRegression.